bokehjs/src/lib/models/tiles/tile_source.ts

```diff
@@ -20,7 +20,17 @@ export namespace TileSource {
     initial_resolution: number
   }
 
-  export interface Props extends Model.Props {}
+  export interface Props extends Model.Props {
+    url: p.Property<string>
+    tile_size: p.Property<number>
+    max_zoom: p.Property<number>
+    min_zoom: p.Property<number>
+    extra_url_vars: p.Property<{[key: string]: string}>
+    attribution: p.Property<string>
+    x_origin_offset: p.Property<number>
+    y_origin_offset: p.Property<number>
+    initial_resolution: p.Property<number>
+  }
```

Comment:
> Let me know if my assumption that attributes need to be properties to trigger model change events is wrong.

Yes and no. If you want automation, then yes. You may try to use `setv()` directly, but that will fail, if "properties" weren't defined. You can, however, emit change signal manually if you want (e.g. from a setter). I'm writing this in the context of the above change, because it isn't necessary in this PR (though useful in general). From runtime perspective only `this.define(...)` matters. Also note that `define()` is untyped currently, so you can put whatever you want there (which obviously I don't recommend).

Reply:
Thanks for the clarification, I think in this case it is needed because we want the model to be responsive to any change events, no matter how they were triggered.

Reply:
Sure, the default is to use properties, especially as the part of the API. However, I just want to make sure that it's understood that the code above doesn't have any run time implications. Any interfaces or types (or anything else that is part of TypeScript and not JavaScript) are fully erased at compilation time. The only thing that matters at runtime is `define()`, and that was fully specified already. But as I said, it doesn't harm to add this code (in fact I have a WIP PR that fills all those in), just I want to make sure that be benefits and consequences of such change are fully understood.

Reply:
Ah okay, I did misunderstand you initially then. Thanks for clarifying, the only thing I'm missing is why attributes and properties seem to be duplicating the same type information. Wouldn't it be sufficient to declare them once as either attributes or as properties?

Reply:
> Wouldn't it be sufficient to declare them once as either attributes or as properties?

Yes (\*), however, limitations of the language when we started the transition didn't allow for this. Currently I can use mapped and conditional types, to generate attributes from properties (this direction, because properties carry more information than attributes). I have a WIP PR that fixes that (after layout). However, (*) attributes and properties may, and in fact do in case of specs, carry different types, or even input (constructor) and output (model interface) types may differ. Altogether this can get pretty tricky, but with some type-level magic it can be handled without duplication.

Reply:
> Wouldn't it be sufficient to declare them once as either attributes or as properties?

Just to add more color: the duplication is entirely because we developed a properties system in coffee script, if we had had the current version of typescript at the time we probably could have done else. Now we have the task of unifying our old sort-of-types with TypieScript's actual language capabilities, and that is an onerous task.

Reply:
Thanks for the background, I really appreciate it.